_posts/2016-06-19-graphql-for-mobile.markdown

```diff
@@ -0,0 +1,99 @@
+
+<!-- more -->
+
+### So what is GraphQL
+
+So, you can get the full explaination on [their website](http://graphql.org). Though I found running through [Learn GraphQL](https://learngraphql.com) to really hammer down how it works as an API consumer. Reading the [introduction blog post](https://facebook.github.io/react/blog/2015/05/01/graphql-introduction.html) can be useful too, I'll be doing a quick TLDR though.
+
+GraphQL is an API middle-layer. It acts as an intermediate layer between multiple front-end clents and multiple back-end APIs. This means it can easily coalesce multiple API calls into a single request, this can be a _massive_ user experience improvement when you have a screen that requires information from multiple sources before you can present anything to a user.
+
+<img src="/images/2016-06-19-graphql-for-iOS-devs/graphQL.svg" width=100%>
+
+As a client, you [send](https://github.com/artsy/eigen/blob/dac7c80b66b600f9a45aaae6095544fe420f0bbc/Artsy/Networking/ARRouter.m#L1011) a "JSON-SQL"-like query structure, which is heirarchical and easy to read:
+
+```json
+{
+  artwork(id: "kimber-berry-as-close-to-magic-as-you-can-get") {
+    id
+    additional_information
+
+    is_price_hidden
+    is_inquireable
+  }
+}
+
+```
+
+> This will search for a [specific artwork](https://www.artsy.net/artwork/kimber-berry-as-close-to-magic-as-you-can-get), sending back the Artwork's `id`, `additional_information`, `is_price_hidden` and `is_inquireable`.
+
+It's important to note here, the data being sent _back_ is only what you ask for. This is not defined on the server as a _short_ or _embedded_ version of a model, but the specific data the client requested. When bandwidth and speed is crucial, this is the other way in which GraphQL improves the end-user experience.
+
+So, that's the two killer features:
+
+1. Coelesce Multiple Network Requests.
+2. Only Send The Data You Want.
+
+This is in stark contrast to existing API concepts, like [HAL](http://stateless.co/hal_specification.html) and [JSON-API](http://jsonapi.org) - both of which are optimised for caching, and rely on "one model, one request" types of API access. E.g. a list of Artworks would actually contain a list of hrefs instead of the model data, and you have to fetch each model as a separate request.
+
+### Using GraphQL
+
+Artsy's GraphQL server is (unsurprisingly) open-source, it's at [artsy/metaphysics](https://github.com/artsy/metaphysics). However, it's not publicly accessible, ([yet?](https://github.com/artsy/metaphysics/issues/279)). One of the coolest things about developing against a GraphQL server is that it has a built-in IDE. I can't show you ours, but I can send you to [Clay Allsop's](http://clayallsopp.com) [GraphQLHub](https://www.graphqlhub.com):
+
+[Here][hub] ( I strongly recommend pausing to open that link in a new window. Press cmd + enter to see the results )
```

Comment:
[Here’s](https://www.graphqlhub.com/playground?query=%23%20Hit%20the%20Play%20button%20above!%0A%23%20Hit%20%22Docs%22%20on%20the%20right%20to%20explore%20the%20API%0A%0A%7B%0A%20%20graphQLHub%0A%20%09reddit%20%7B%0A%20%20%20%20user(username%3A%20%22orta%22)%20%7B%0A%09%09%09...userFields%0A%20%20%20%20%7D%0A%20%20%20%20%0A%20%20%20%20subreddit(name%3A%20%22swift%22)%20%7B%0A%20%20%20%20%20%20newListings(limit%3A%202)%20%7B%0A%20%20%20%20%20%20%20%20title%0A%20%20%20%20%20%20%20%20comments%20%7B%0A%20%20%20%20%20%20%20%20%20%20body%0A%20%20%20%20%20%20%20%20%20%20author%20%7B%20%0A%09%09%09%20%20%20%20%20%20...userFields%0A%20%20%20%20%20%20%20%20%20%20%7D%0A%20%20%20%20%20%20%20%20%7D%0A%20%20%20%20%20%20%7D%0A%20%20%20%20%7D%0A%20%20%7D%0A%7D%0A%0Afragment%20userFields%20on%20RedditUser%20%7B%0A%20%20username%0A%20%20commentKarma%0A%20%20createdISO%0A%7D) an example that refactors the common user fields into a fragment, if that’s something that interests you. I also added some whitespace before the curly brace in: `subreddit(name: "swift") {`